src/main/java/org/broadinstitute/hellbender/tools/dragstr/EstimateDragstrParameters.java

```diff
@@ -0,0 +1,935 @@
+    private String sitesOutput = null;
+
+    private SAMSequenceDictionary dictionary;
+
+    @Override
+    public boolean requiresReference() {
+        return true;
+    }
+
+    @Override
+    public boolean requiresReads() {
+        return true;
+    }
+
+    @Override
+    protected void onStartup() {
+        super.onStartup();
+        hyperParameters.validate();
+        dictionary = getBestAvailableSequenceDictionary();
+        if (runInParallel) {
+            if (threads == 1) {
+                logger.warn("parallel processing was requested but the number of threads was set to 1");
+            }
+        } else if (threads > 1) {
+            runInParallel = true;
+        }
+        if (runInParallel) {
+            if (threads == 0) {
+                logger.info("Running in parallel using the system suggested default thread count: " + Runtime.getRuntime().availableProcessors());
+            } else {
+                logger.info("Running in parallel using the requested number of threads: " + threads);
+            }
+        }
+    }
+
+    @Override
+    public void traverse() {
+        hyperParameters.validate();
+        dictionary = getBestAvailableSequenceDictionary();
+        final List<SAMReadGroupRecord> readGroups = hasReads() ? getHeaderForReads().getReadGroups() : Collections.emptyList();
+        final List<String> readGroupIds = readGroups.stream()
+                .map(SAMReadGroupRecord::getId)
+                .collect(Collectors.toList());
+        final List<String> sampleNames = readGroups.stream()
+                .map(SAMReadGroupRecord::getSample)
+                .distinct().collect(Collectors.toList());
+        final Optional<String> sampleName = resolveSampleName(sampleNames);
+
+        try (final PrintWriter sitesOutputWriter = openSitesOutputWriter(sitesOutput);
+             final STRTableFile strTable = STRTableFile.open(strTablePath)) {
+
+            checkSequenceDictionaryCompatibility(dictionary, strTable.dictionary());
+            final StratifiedDragstrLocusCases allSites;
+            final List<SimpleInterval> intervals = getTraversalIntervals();
+
+            runInParallel |= threads > 1;
+            if (runInParallel) {
+                if (threads == 1) {
+                    logger.warn("parallel processing was requested but the number of threads was set to 1");
+                }
+                allSites = collectCaseStatsParallel(intervals, shardSize, strTable);
+            } else {
+                allSites = collectCaseStatsSequencial(intervals, strTable);
+            }
+            logSiteCounts(allSites, "all loci/cases");
+            final StratifiedDragstrLocusCases downSampledSites = downSample(allSites, strTable, sitesOutputWriter);
+            logSiteCounts(downSampledSites, "all downsampled (kept) loci/cases");
```

Comment:
True that does not make much sense but that is what DRAGEN does at least v3.4, I think the newer version do all the filtering first.... This is actually a bit worse that you depict it as depth is considered twice. First before decimation we remove sites that have depth == 0, then decimate the again filter out depth < 10 and some other criteria; depth is looked at twice and the first one is weaker than the second one. 

Why? you would ask. My guess is that it is due to progressive refactoring of this process... if the original was any similar to the way the matlab scripts work (and the first version that I submitted in the PR) then I can see why it would end up this way as the subset was been synthesized adding more sites at random rather than decimated all the sites and then check. I bet that before forming the pileup there would just look at the number of reads available (depth > 0) and only then actually build the pileup and do the filtering

Now going back to the present...
So DRAGEN uses the first decimation in ComposeSTRTableFile to reduce the number of sites to something that they feel comfortable to handle with their very parallel FPGAs. And the second stage is just kind a "brute-force" relaying in their superior and fast hardware. We pay the price in GATK because we don't have that. 


There is even a better strategy that we could implement here that would give the same result, but I didn't know until the very end because just until recently didn't exactly know what DRAGEN was doing in what order and as you can see it wasn't that intuitive:

You first decimated to at least 4096 (usually something between 8192  and 4096) and only then access their alignments and filter. If more than 4096 pass then you decimate back down one step so that you have less or equal to 4096, you are done. If there are less 4096 you need to un-decimate in bit down (give you roughly another 8192 cases in the first interaction) and repeat until you get more than 4096. This is kind of wasteful still but that waste is needed to match the DRAGEN approach

Reply:
This in turn may remove the need of parallelism so I think that should be a part of that ticket/issue but only into we have implemented their newer approach 3.5 3.6+ and we match what they do there using this rather inefficient approach.